app/controllers/application_controller.rb

```diff
@@ -1,10 +1,79 @@
-require './config/environment'
-
 class ApplicationController < Sinatra::Base
 
   configure do
     set :public_folder, 'public'
     set :views, 'app/views'
+    enable :sessions
+    set :session_secret, "password_security"
+  end
+
+  get '/' do
+    erb :index
+  end
+
+  get '/signup' do
+    if !logged_in?
+      erb :signup
+    else
+      redirect '/tweets'
+    end
+  end
+
+  post '/signup' do
+    redirect '/signup' unless user_credentialed?
+    create_user
+    session[:user_id] = @user.id
+    redirect '/tweets'
+  end
+
+  get '/login' do
+    if !logged_in?
+      erb :login
+    else
+      redirect '/tweets'
+    end
+  end
+
+  post '/login' do
+    find_user && authenticate_user
+    redirect '/login' unless user_exists?
+    session[:user_id] = @user.id
+    redirect '/tweets'
+  end
+
+  get '/logout' do
+    session.clear if logged_in?
+    redirect '/login'
+  end
+
+  helpers do
+    def current_user
+      @user = User.find(session[:user_id])
```

Comment:
Because in the lifespan of a request, you might call `current_user` multiple times, we'd prefer this method to only hit the database once and then 'cache' the user instance into an instance variable that could be returned on any subsequent call to `current_user`.

```ruby
def current_user
  @current_user ||= User.find(session[:id) if session[:id]
end
```

This pattern of conditionally setting an instance variable or returning it if it's already defined using the `||=` (or-equals) operator is called Memoization. ([read more](http://gavinmiller.io/2013/basics-of-ruby-memoization/) and [here](http://www.justinweiss.com/articles/4-simple-memoization-patterns-in-ruby-and-one-gem/))